inference.py

```diff
@@ -27,6 +27,117 @@
 from osu_diffusion.config import DiffusionTrainConfig
 
 
+def get_directory_from_file_path(file_path):
+    """
+    Extract directory path from a file path.
+
+    Args:
+        file_path (str): Path to a file
+
+    Returns:
+        str: Directory path, or empty string if invalid
+    """
+    if not file_path or file_path.strip() == "":
+        return ""
+
+    try:
+        path = Path(file_path)
+        return str(path.parent)
+    except Exception:
+        return ""
+
+
+def validate_audio_file_exists(audio_path):
+    """
+    Check if audio file exists at the given path.
+
+    Args:
+        audio_path (str): Path to audio file
+
+    Returns:
+        dict: {'exists': bool, 'error': str or None}
+    """
+    if not audio_path or audio_path.strip() == "":
+        return {'exists': False, 'error': None}
+
+    try:
+        path = Path(audio_path)
+        if path.exists() and path.is_file():
+            return {'exists': True, 'error': None}
+        else:
+            return {'exists': False, 'error': f"Audio file not found: {audio_path}"}
+    except Exception as e:
+        return {'exists': False, 'error': f"Error checking audio file: {str(e)}"}
+
+
+def get_autofill_paths_from_beatmap(beatmap_path):
```

Comment:
this can just be a simple function that returns the audio_path for a beatmap_path if it exists and return '' if it doesn't exist. No need to overcomplicate things.